### PR TITLE
tomatsu-testing

### PR DESCRIPTION
**Pull request by ltomatsu**

### CODE REVIEW

**Comment by ltomatsu** [comment]
I originally pull requested using the github website, but I found the CLI much more simpler. Please let me know which one you prefer and how I can improve!

File: config.py [hunk 1]
```diff
@@ -34,3 +34,7 @@ def set_config_file(cfg_file=None):
             raise IOError("No valid configuration files found.")
 
     return CONFIG
+
+# If the index name of USER_COLUMN or PROJECT_COLUMN changes, change the name of these variables below
+user_column = 'Email Address'
```
Comment (kamfonik) [inline config.py]: I think I was unclear, this should be in the settings.ini file, not config.py.

File: addusers.py [hunk 2]
```diff
@@ -46,7 +46,7 @@
 from moc_utils import get_absolute_path, select_rows
 from quotas import QuotaManager
 from setpass import SetpassClient, random_password
-from config import set_config_file
+from config import set_config_file, user_column
```
Comment (kamfonik) [inline addusers.py]: Instead of importing this from config.py, it should be read in from the settings.ini file using ConfigParser.  If you look at line 354-358 where the admin credentials are loaded, you can see how to do this.

For now, you can assume the `config` variable from `__main__` will be inherited in the functions.  There are some other functions that also do this at the moment, it will be changed later when the configuration loading is streamlined.

File: addusers.py [hunk 3]
```diff
@@ -293,7 +293,12 @@ def parse_rows(rows, select_user=None):
                     row=idx, name=entry[15],
                     contact_name=user.first_name + " " + user.last_name,
                     contact_email=user.email,
-                    description='org: {}, contact_name: {}, contact_email: {}, description: {}'.format(entry[8], user.first_name + " " + user.last_name, user.email, entry[16]),
+                    description='org: {},contact_name: {},contact_email: {},\
+                    description: {}'.format(
```
Comment (kamfonik) [inline addusers.py]: This is going to put many spaces in your string.  You can put the `\` outside the quotes:    
```
my_long_string = "This is the first line of a long string. " \
                 "This is the second line."
```
This is another way to write the same thing:   
 ```
my_long_string = ("This is the first line of a long string. "
                   "This is the second line.")
```
Both of these will print `This is the first line of a long string.  This is the second line.`  You can do .format() at the end of either version and either is fine in terms of PEP8 etc.  Although it's good to be consistent within a file about which one to use, the second version is used elsewhere in this file.

File: addusers.py [hunk 4]
```diff
@@ -356,6 +361,7 @@ def parse_rows(rows, select_user=None):
     admin_project = config.get('auth', 'admin_project')
     auth_url = config.get('auth', 'auth_url')
     nova_version = config.get('nova', 'version')
+    user_column = config.get('header_column', 'user_column')
```
Comment (kamfonik) [inline addusers.py]: Please also update `example_settings.ini` with the example section and these settings.

File: addusers.py [hunk 5]
```diff
@@ -212,7 +212,7 @@ def parse_rows(rows, select_user=None):
     """
     # Column index in the Google Sheet for username
     # This may need to be updated if question order on the form is changed
-    USER_COLUMN = 4
+    USER_COLUMN = rows[0].index(user_column)
```
Comment (kamfonik) [inline addusers.py]: This is nitpicky, but it would be better to do the `config.get()` step right here, to minimize how much we are assuming about the environment.  I've been allowing functions in this script to assume they will inherit the `config` variable, but I'd rather not assume too much else.

It would also be good to change `USER_COLUMN` to `user_column` since it's no longer a hard-coded constant, and use some other name ( `user_column_header` or something) for the string value.

Both of the above also apply to PROJECT_COLUMN in the other script.

Comment (ltomatsu) [inline addusers.py]: Got it!

File: addusers.py (hunk 3 above)
Comment (ltomatsu) [inline addusers.py]: I find the second option easier

File: addusers.py (hunk 4 above)
Comment (ltomatsu) [inline addusers.py]: Will do!

File: addusers.py [hunk 6]
```diff
@@ -201,7 +203,106 @@ def exists_in_keystone(check_thing, keystone_things):
         return None
 
 
-def parse_rows(rows, select_user=None):
+def remove_duplicates(rows, UC, PC, EC):
```
Comment (kamfonik) [inline addusers.py]: What is this for?  Why do we expect duplicates in `rows`?

File: addusers.py [hunk 7]
```diff
@@ -201,7 +203,106 @@ def exists_in_keystone(check_thing, keystone_things):
...
+            if flag:
+                output.append(rows[i])
+    return output
+
+
+def create_new_users_rows(rows):
+    output = []
+    for entry in rows:
+        if entry[7] == 'No':
+            user = [entry[4], entry[5], entry[6], entry[8],
```
Comment (kamfonik) [inline addusers.py]: There should be a more readable way to do this with the User class, that long list of `entry[x]` will be obnoxious to maintain.  It's also not very efficient to iterate again here - you could do this work at the end of the `create_user()` function.

Same thing goes for `create_new_projects_rows()` below

File: addusers.py [hunk 8]
```diff
@@ -201,7 +203,106 @@ def exists_in_keystone(check_thing, keystone_things):
...
+                       entry[13], entry[16], entry[17]]
+            output.append(project)
+    return output
+
+
+def manual_testing(header):
+    """Testing changes to the code that doesn't involve the form
+    """
+    print "...TESTING..."
+    header_option = raw_input("Create own header?     [y/n] ")
```
Comment (kamfonik) [inline addusers.py]: I'm not sure we need a choice to create our own header or not.  The rows read in from the spreadsheet will always have a header, so tests should always include one.

File: addusers.py [hunk 9]
```diff
@@ -201,7 +203,106 @@ def exists_in_keystone(check_thing, keystone_things):
...
+        if index_flag == 'y':
+            index_length -= 4
+        entry = [0] * len(header)
+        for i in range(len(entry)):
+            entry[i] = raw_input(header[i] + ": ")
+        tests.append(entry)
+    return tests
+
+
+def manual_fix(rows):
```
Comment (kamfonik) [inline addusers.py]: I'm confused about what makes an input incorrect?

File: addusers.py [hunk 10]
```diff
@@ -201,7 +203,106 @@ def exists_in_keystone(check_thing, keystone_things):
...
+    for entry in rows:
+        if entry[15] != "":
+            project = [entry[4], entry[5], entry[6], entry[8],
+                       entry[9], entry[10], entry[11], entry[12],
+                       entry[13], entry[16], entry[17]]
+            output.append(project)
+    return output
+
+
+def manual_testing(header):
```
Comment (kamfonik) [inline addusers.py]: So, this is solving a completely different problem than the original PR.  In general, each PR should deal with a single topic to make things easy for reviewers.

Can you please submit a different PR for each bug you are fixing?

File: config.py [hunk 11]
```diff
@@ -34,3 +34,14 @@ def set_config_file(cfg_file=None):
             raise IOError("No valid configuration files found.")
 
     return CONFIG
+
+
+"""Oslo_log setup"""
+from oslo_config import cfg
```
Comment (kamfonik) [inline config.py]: Imports should go at the top of the file - I am surprised flake8 doesn't complain here?

File: addusers.py (hunk 10 above)
Comment (ltomatsu) [inline addusers.py]: I will submit a different PR for each bug!

Comment (ltomatsu) [inline addusers.py]: I will submit a different PR!

File: addusers.py (hunk 8 above)
Comment (ltomatsu) [inline addusers.py]: okay

File: addusers.py (hunk 9 above)
Comment (ltomatsu) [inline addusers.py]: An incorrect input could be a typo that the user made.
For instance, the incorrect input is '1234@email.com', but the correct input is '5678@email.com'.

File: config.py (hunk 11 above)
Comment (ltomatsu) [inline config.py]: flake8 did not complain here but I will add it to the beginning

File: addusers.py (hunk 7 above)
Comment (ltomatsu) [inline addusers.py]: I am using the config to do this better

File: addusers.py (hunk 6 above)
Comment (ltomatsu) [inline addusers.py]: Before we call the delete_rows() function by spreadsheet function, I checked if there are any duplicates in rows to avoid bug #23.

**Comment by ltomatsu** [comment]
I will be creating a new PR for each issue I will work on.

**Comment by ltomatsu** [comment]
I backed up everything in my external drive. I decided to delete my forked repo to organize things. This pull request may be deleted as well.
